### ISSUE TITLE
[FEAT]: Support GitHub repository custom properties

### ISSUE THREAD

**Issue by yu-iskw** [integrations/terraform-provider-github#1956]
### Describe the need

GitHub just announced the feature of GitHub repository custom properties. We can use the information of custom properties to apply rules to repository in GitHub organization(s). It would be good to support the feature in the terraform provider as well. By doing that, we will be able to manage custom properties and rule sets in terraform.

https://github.blog/changelog/2023-10-12-github-repository-custom-properties-beta/

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by kfcampbell** [comment]
Unfortunately, this feature is not released on the API yet. Once support is added in the API, we can implement it in the provider.

**Comment by usmonster** [comment]
Looks like [a REST API](https://docs.github.com/en/rest/repos/custom-properties?apiVersion=2022-11-28) is in public beta now. Is that enough to get started? Thanks!

**Comment by kfcampbell** [comment]
Ideally there would be more than just a GET. We could make a data source with the existing endpoint, but not a full resource.

**Comment by laughedelic** [comment]
There are more endpoints here (incl. creation): https://docs.github.com/en/rest/orgs/custom-properties?apiVersion=2022-11-28

**Comment by kfcampbell** [comment]
Ahh my mistake, thanks for correcting me!

**Comment by roeishuster** [comment]
Hey, are you guys plan on adding this in an upcoming provider update?

**Comment by kfcampbell** [comment]
GitHub's SDK team unfortunately does not have the bandwidth to take this on ourselves, though we're very receptive of PRs!

**Comment by alileza** [comment]
@kfcampbell I'll make an attempt to PR soon, just drafted init

**Comment by nayuta** [comment]
Hi, I also need to support custom properties for repositories. However, it needs to update go-github. So, I created the issue: https://github.com/integrations/terraform-provider-github/issues/2187

**Comment by kfcampbell** [comment]
It's worth noting that we (GitHub's SDK team) are working on a [Go SDK](https://github.com/octokit/go-sdk) generated from GitHub's OpenAPI specification. It includes all routes published in the GitHub API, so coverage doesn't need to be added manually. I've been looking for an excuse to add it to the project; this could be it.

**Comment by Simon-Boyer** [comment]
I have a client that will need this very soon. If I can help fast-track this by contributing I can put some hours in this. I see there is already a PR opened and some discussions around the Go SDK, is there anything I can do here?

**Comment by connor-miller-kr** [comment]
Any update on this feature?

**Comment by Simon-Boyer** [comment]
> Any update on this feature?

I wouldnt mind contributing for this feature as soon as the go-github client is bumped to v60. But that PR looks a bit stuck...

**Comment by larmitage-bjss** [comment]
Now the go-github PR has been merged is there potential for any progress on this feature soon? Thanks!

**Comment by Hulidex** [comment]
Any update so far? Thank you!!

**Comment by felixlut** [comment]
#2316 is blocked by #2443

**Comment by philthethrill99** [comment]
> #2316 is blocked by #2443

Now that #2443 is merged, do you have an ETA on this feature request?

Thanks

**Comment by chapurlatn** [comment]
Hello,
I'd like to be able to manage custom properties during repo creation as well.
The API allows to provide properties just like it allows to provide Topics when creating a repository.
Do you think we could imagine adding the feature in your PR to implement a full custom property support or do you prefer me to open an additional pull request to handle it?

**Comment by felixlut** [comment]
> Hello, I'd like to be able to manage custom properties during repo creation as well. The API allows to provide properties just like it allows to provide Topics when creating a repository. Do you think we could imagine adding the feature in your PR to implement a full custom property support or do you prefer me to open an additional pull request to handle it?

Another PR. #2316 is big enough as it is

**Comment by hashildy** [comment]
Any update on this feature request? Looks like progress has been made on #2316 but stalled for the past few weeks...

**Comment by meridius** [comment]
For anyone interested... To **set** custom properties on **repository** using **GitHub App**, Read-**Write** permission is needed on **organization**; yes, really.

**Comment by FedeBerbara** [comment]
> For anyone interested... To **set** custom properties on **repository** using **GitHub App**, Read-**Write** permission is needed on **organization**; yes, really.
> 
> The missing permission on the App will get you:
> 
> ```
> │ Error: PATCH https://api.github.com/repos/{owner}/{repo}/properties/values: 422 Actor doesn't have permissions to edit properties [{propName}] []
> │
> │   with github_repository_custom_property.this["{propName}"],
> │   on custom_properties.tf line 8, in resource "github_repository_custom_property" "this":
> │    8: resource "github_repository_custom_property" "this" {
> ```

Hey @meridius how are you ? Which permission of organization I need to set on my GitHub App then ? Do you know why read-write on the custom propertie isn't enough ? Is a bug ?

**Comment by meridius** [comment]
Hey @FedeBerbara, as I wrote - your GH App needs to have R/W for custom props on Org.
